proposals/0000-overloaded-bracket.rst

```diff
@@ -0,0 +1,394 @@
+untyped syntax, the implementation of this is natural.
+
+Motivation 2: Effectful Code Generation
+---------------------------------------
+
+Jamie Willis provides additional motivation for the generalisation of the
+quotation bracket. Whilst writing multi-stage programs it is almost inevitable
+that you will need to perform effects whilst doing the code generation.
+For example, this is from the abstract of Kameyama, Kiselyov and Shan's (2014) Combinators
+for impure yet hygienic code generation:
+
+    Code generation is the leading approach to making high-performance software reusable. Effects are indispensable in code generators, whether to report failures or to insert let-statements and if- guards.
+
+To be precise, in his parser combinator library which is implemented using
+typed template haskell the following effects are necessary.
+
+1. Use the ``Reader`` monad in order to carry around an environment.
+2. Use a let-insertion effect to automatically insert lets to avoid code duplication.
+3. Use an exception monad to automatically insert missing dependencies for mutually recursive code generation.
+
+Even using the simple ``Reader`` monad is awkward at the moment::
+
+   generateLoop :: String -> Reader CodeMap (Q Exp)
+   generateLoop name = ask (\codeMap -> [|
+     let loopyCode x =
+       $(runReader loopBody (Map.insert name [|loopyCode|] codeMap))
+     in loopyCode ...
+     |])
+
+The effect must be explicitly run in each splice. For state or other more complicated
+effects this approach doesn't work. With the proposal we would hope to write something like::
+
+   generateLoop :: (MonadReader CodeMap m, Quote m) => String -> m Exp
+   generateLoop name = [|
+     let loopyCode x =
+       $(local (Map.insert name [|loopyCode|]) loopBody)
+     in loopyCode ...
+     |]
+
+and directly use the ``local`` function inside the nested splice just like normal
+monadic programming.
+
+
+Proposed Change Specification
+-----------------------------
+
+The goal of the changes is for an expression ``e : T`` to give the
+representation ``[| e |] : Quote m => m Exp``. Several steps are necessary to
+make this change possible.
+
+1. Define the interface for ``Quote``::
+
+      class Applicative m => Quote m where
+         newName :: String -> m Name
+
+   These are all the operations which are necessary to build the representation
+   of expressions.
+
+2. Generalise all the combinators which build syntax in ``Language.Haskell.TH.Lib``.
+   Due to an `audit <https://github.com/ghc-proposals/ghc-proposals/issues/211#issuecomment-472092412>`_
+   conducted by Richard, it was found that the only effect from
+   ``Q`` which was used is the ``newName`` function which generates a fresh
+   name. All the other combinators can be defined using the ``Applicative``
+   operations.
+
+
+3. Generalise the ``Lift`` type class::
+
+      class Lift a where
+         lift :: Quote m => a -> m Exp
```

Comment:
I'm wondering if there should be a comparison with

```haskell
class Quote m => Lift m a where
  list :: Quote m => a -> m Exp
```

This would allow for more flexible `Lift` instances, which would work in some quotations but not other. On the other hand, we may actively not care about such instances, and maybe type inference wouldn't be as good?

Still, it probably makes sense to mention it in the alternatives.

Reply:
> #195 certainly can be so changed, but I worry that the change is against it's spirit if not it's letter. I see #195 as an attempt to hide "implementation details" that also mildly interfere with invariants. But if we expose a monad parameter like that, we're saying the monad is a rightful part of the interface rather than implementation detail, in which case the `Code` datatype (IMO) appears like an odd attempt to ram together two independent axes of variation into one type, very much unlike the way we usually deal with monads.

@Ericson2314 I think the primary concern was with people using effects from ``Q``, using effects during code generation is a completely natural thing to do in both typed and untyped metaprogramming.

Reply:
This is quite a big breaking change which the proposal has made quite an effort to avoid. It would be better to do this later in an wholesale change to `Lift` rather than incrementally.